### PR TITLE
Qt/PropertiesDialog: Add "Game Config" tab

### PR DESCRIPTION
**Pull request by spycrab**
Screenshot

### CODE REVIEW

**Comment by spycrab** [comment]
@MayImilae @JMC47

**Comment by BhaaLseN** [comment]
Any thoughts on making those determinism string values `constexpr` as well?

**Comment by spycrab** [comment]
Sounds reasonable.

File: Source/Core/DolphinQt2/Config/GameConfigWidget.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+private:
+  void CreateWidgets();
+  void ConnectWidgets();
+  void LoadSettings();
+  void SaveSettings();
+
+  void EditUserConfig();
+  void ViewDefaultConfig();
+
+  void LoadCheckBox(QCheckBox* checkbox, const std::string&& section, const std::string&& key);
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.h]: Do const rvalues really provide a benefit here over const references?

File: Source/Core/DolphinQt2/Config/GameConfigWidget.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#pragma once
+
+#include <string>
+
+#include <QSpinBox>
+#include <QString>
+#include <QWidget>
+
+#include "Common/IniFile.h"
+#include "DolphinQt2/GameList/GameFile.h"
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.h]: As far as I'm aware, GameFile can be forward declared.

File: Source/Core/DolphinQt2/Config/GameConfigWidget.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <string>
+
+#include <QSpinBox>
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.h]: QSpinBox is already forward declared further down.

**Comment by spycrab** [comment]
@lioncash All fixed now.

**Comment by MayImilae** [comment]
Ok, took a look at this window! I'll start with macOS.

screen shot 2018-03-04 at 7 18 16 pm

This is something of a common problem with Qt right now in macOS - everything is **HUGE!!** The spacing is all wrong and the window is too big for my poor laptop's screen. Considering none of these spacing problems exist on the linux screenshots, I'm assuming this is just a macOS issue that needs to be ironed out. Desperately, it's bad.

...or maybe it's on windows too, ok.

spacing windows

The spacing is all wrong here too. Not as bad, I mean, it isn't suuuuper wide as well as super tall, but it's pretty overly spaced. Note that unlike macOS, I can make it smaller, but by default it is at the size shown above. Even when shrunken to the minimum though, there's still wasted space. Around the text at the top and such, see below. Note that if I close the window and reopen it it is at the bigger size above.

spacing windowsshrunken

This is not related to this PR so I won't go into detail here, but there are spacing problems elsewhere too. Spacing issues are kind of a big problem with Qt right now, with tons of things in Windows and macOS (especially) just having waisted space for no apparent reason. Yay platform specific issues! We'll need to get that ironed out before we make the switch. I'll talk with you over IRC when you're available, I've written up a list! :3

Since the spacing problems are so widespread, if you want to just merge this and fix all the spacing issues together, I'd be fine with that. 

Anyway, aside from the spacing issues... I'm not what the wii widescreen option is doing here? I mean, sure it would have been useful years ago, but all of that was handled with the heuristic and gameinis now. Wii games that need to be 4:3 only, like virtual console titles, have that already set in the gameinis dolphin ships with. So there isn't a reason to promote that for easier user access. Why did you add it?

Also the buttons at the bottom of the game config tab (edit user config and view default config) are too tight could use more space around the text. I would see if like that's on macOS too, but I can't see them!

**Comment by spycrab** [comment]
> Anyway, aside from the spacing issues... I'm not what the wii widescreen option is doing here? I mean, sure it would have been useful years ago, but all of that was handled with the heuristic and gameinis now. Wii games that need to be 4:3 only, like virtual console titles, have that already set in the gameinis dolphin ships with. So there isn't a reason to promote that for easier user access. Why did you add it?

Removed.

> Also the buttons at the bottom of the game config tab (edit user config and view default config) are too tight could use more space around the text. I would see if like that's on macOS too, but I can't see them!

Fixed.

I will address the other "general" spacing issues in another PR.

**Comment by spycrab** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/DolphinQt2/Config/GameConfigWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+}
+
+void GameConfigWidget::LoadCheckBox(QCheckBox* checkbox, const std::string& section,
+                                    const std::string& key)
+{
+  bool checked;
+
+  if (m_gameini_local.GetOrCreateSection(section)->Get(key, &checked))
+    return checkbox->setCheckState(checked ? Qt::Checked : Qt::Unchecked);
+  else if (m_gameini_default.GetOrCreateSection(section)->Get(key, &checked))
```
Comment (leoetlino) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.cpp]: You don't need an `else` after a `return`.

File: Source/Core/DolphinQt2/Config/GameConfigWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+      m_gameini_local.GetOrCreateSection(section)->Set(key, checked);
+    else
+      m_gameini_local.DeleteKey(section, key);
+
+    return;
+  }
+
+  m_gameini_local.GetOrCreateSection(section)->Set(key, checked);
+
+  return;
```
Comment (leoetlino) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.cpp]: This return doesn't seem to serve any purpose, could you remove it?

File: Source/Core/DolphinQt2/Config/GameConfigWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+  LoadCheckBox(m_sync_gpu, "Core", "SyncGPU");
+  LoadCheckBox(m_enable_fast_disc, "Core", "FastDiscSpeed");
+  LoadCheckBox(m_use_dsp_hle, "Core", "DSPHLE");
+
+  std::string determinism_mode;
+  if (!m_gameini_local.GetIfExists("Core", "GPUDeterminismMode", &determinism_mode))
+    if (!m_gameini_default.GetIfExists("Core", "GPUDeterminismMode", &determinism_mode))
+      determinism_mode = DETERMINISM_NOT_SET_STRING;
+
+  if (determinism_mode == DETERMINISM_AUTO_STRING)
```
Comment (leoetlino) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.cpp]: IMO, a switch statement would be better here.

File: Source/Core/DolphinQt2/Config/GameConfigWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+  {
+    m_deterministic_dual_core->setCurrentIndex(DETERMINISM_FAKE_COMPLETION_INDEX);
+  }
+  else
+  {
+    m_deterministic_dual_core->setCurrentIndex(DETERMINISM_NOT_SET_INDEX);
+  }
+
+  // Stereoscopy
+  int depth_percentage = 100;
```
Comment (leoetlino) [inline Source/Core/DolphinQt2/Config/GameConfigWidget.cpp]: There is quite a lot of code duplication here (there are 3 calls to `setValue` for one variable). I'd just do:

```
int depth_percentage = 100;
m_gameini_default.GetIfExists(...);
m_gameini_local.GetIfExists(...);
m_depth_slider->setValue(depth_percentage);
```

File: Source/Core/DolphinQt2/Config/GameConfigWidget.cpp (hunk 6 above)
Comment (spycrab) [reply Source/Core/DolphinQt2/Config/GameConfigWidget.cpp]: A switch on a string?

Comment (leoetlino) [reply Source/Core/DolphinQt2/Config/GameConfigWidget.cpp]: Oh whoops, disregard what I said.

**Pull request merged**